tools/shared/src/main/scala/org/scalajs/core/tools/io/IRFileCache.scala

```diff
@@ -0,0 +1,314 @@
+
+    /** Last version we have been updated with.
+     *  May only be written under synchronization, except this is a tombstone
+     */
+    @volatile
+    private[this] var _version: Option[String] = None
+
+    /** Files in this [[PersistedFiles]]
+     *  May only be written under synchronization, except this is a tombstone
+     */
+    @volatile
+    private[this] var _files: Seq[VirtualRelativeIRFile] = null
+
+    def files: Seq[VirtualRelativeIRFile] = _files
+
+    /** Try to reference this block of files.
+     *  @return true if referencing succeeded, false if this is a tombstone
+     */
+    @tailrec
+    final def reference(): Boolean = {
+      val refs = _references.get
+
+      if (refs == -1) {
+        // we are a tombstone, help cleaning up and bail out
+        cleanup()
+        false
+      } else {
+        // try to increase ref count
+        if (_references.compareAndSet(refs, refs + 1)) true // done
+        else reference() // something changed, try again
+      }
+    }
+
+    /** Dereference this file.
+     *
+     *  If there are no references any more, turn this [[PersistedFiles]] into
+     *  a tombstone and remove it from the cache.
+     */
+    final def dereference(): Unit = {
+      val refs = _references.decrementAndGet()
+      assert(refs >= 0, "Dereferencing an unreferenced file")
+
+      /* If we have 0 references, try to become a tombstone. We could be
+       * referenced again in a race. In this case, don't do anything.
+       */
+      if (refs == 0 && _references.compareAndSet(0, -1)) {
+        cleanup()
+      }
+    }
+
+    /** Clean up, after becoming a tombstone */
+    private def cleanup(): Unit = {
+      /* We need to verify our own identity. Otherwise we might mess with a new,
+       * clean state after an exception (if we are pre-exception).
+       */
+      globalCache.remove(path, this)
+      // aggressively dereference stuff for GC
+      _version = null
+      _files = null
+    }
+
+    /** Updates this file with the given [[IRContainer]].
+     *
+     *  May only be called by a thread, if it holds a reference to this file.
+     */
+    def update(file: IRContainer): Unit = {
+      assert(_references.get > 0, "Updating an unreferenced file")
+      assert(file.path == path, s"Path mismatch: $path, ${file.path}")
+
+      // Helper to ensure v is stable during check
+      @inline
+      def upToDate(v: Option[String]) = v.isDefined && v == file.version
+
+      if (upToDate(_version)) {
+        // yeepeeh, nothing to do
+        statsReused.incrementAndGet()
+      } else {
+        // We need to update this. We synchronize
+        synchronized {
+          if (upToDate(_version)) {
+            // someone else had the same idea and did our work
+            statsReused.incrementAndGet()
+          } else {
+            statsInvalidated.incrementAndGet()
+            _files = extractIRFiles(file).map(new PersistentIRFile(_))
```

Comment:
Sounds reasonable. The inside of jars are unlikely to change between reloads anyway (except with `exportJars := true`, but that's pretty rare).